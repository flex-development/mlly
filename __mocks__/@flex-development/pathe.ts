/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export default {
  basename: vi.fn(actual.basename),
  changeExt: vi.fn(actual.changeExt),
  dirname: vi.fn(actual.dirname),
  extname: vi.fn(actual.extname),
  format: vi.fn(actual.format),
  isAbsolute: vi.fn(actual.isAbsolute),
  join: vi.fn(actual.join),
  parse: vi.fn(actual.parse),
  relative: vi.fn(actual.relative),
  removeExt: vi.fn(actual.removeExt),
  resolve: vi.fn(actual.resolve)
}
