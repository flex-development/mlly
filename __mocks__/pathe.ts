/**
 * @file Mocks - pathe
 * @module mocks/pathe
 * @see https://github.com/unjs/pathe
 */

/**
 * [`pathe`][1] module type.
 *
 * [1]: https://github.com/unjs/pathe
 */
type Actual = typeof import('pathe')

/**
 * `pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const isAbsolute = vi.fn(actual.isAbsolute)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)
