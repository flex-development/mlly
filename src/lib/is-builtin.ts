/**
 * @file isBuiltin
 * @module mlly/lib/isBuiltin
 */

import { builtinModules } from 'node:module'

/**
 * Checks if `name` is a [builtin module][1].
 *
 * [1]: https://nodejs.org/api/esm.html#builtin-modules
 *
 * @example
 *  isBuiltin('@flex-development/mlly') // false
 * @example
 *  isBuiltin('node:fs') // true
 * @example
 *  isBuiltin('fs') // true
 *
 * @param {string} name - Module name to check
 * @return {boolean} `true` if `name` is builtin module, `false` otherwise
 */
const isBuiltin = (name: string): boolean => {
  return builtinModules.flatMap(m => [m, 'node:' + m]).includes(name)
}

export default isBuiltin
