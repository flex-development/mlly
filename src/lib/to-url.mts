/**
 * @file toUrl
 * @module mlly/lib/toUrl
 */

import canParseUrl from '#lib/can-parse-url'
import { isBuiltin } from '@flex-development/is-builtin'
import type { ModuleId } from '@flex-development/mlly'
import pathe from '@flex-development/pathe'

/**
 * Convert `id` to a {@linkcode URL}.
 *
 * > ðŸ‘‰ **Note**: If `id` cannot be parsed as a `URL` (in conjunction with
 * > `parent`) and is also not a [builtin module][builtin-module], it will be
 * > assumed to be a path and converted to a [`file:` URL][file-url].
 *
 * [builtin-module]: https://nodejs.org/api/esm.html#builtin-modules
 * [file-url]: https://nodejs.org/api/esm.html#file-urls
 *
 * @param {ModuleId} id
 *  The module id to convert
 * @param {ModuleId | null | undefined} [parent]
 *  Base URL to resolve against if `id` is not absolute
 * @return {URL}
 *  New URL
 */
function toUrl(id: ModuleId, parent?: ModuleId | null | undefined): URL {
  return canParseUrl(id, parent)
    ? new URL(id, parent ?? undefined)
    : isBuiltin(id)
    ? new URL('node:' + String(id))
    : pathe.pathToFileURL(String(id))
}

export default toUrl
