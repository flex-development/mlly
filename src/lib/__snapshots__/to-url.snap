// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:lib/toUrl > should return \`id\` as \`URL\` (0) 1`] = `"node:os"`;

exports[`unit:lib/toUrl > should return \`id\` as \`URL\` (1) 1`] = `file://\${process.cwd()}/path`;

exports[`unit:lib/toUrl > should return \`id\` as \`URL\` (2) 1`] = `file://\${process.cwd()}/src/lib/__tests__/to-url.spec.mts`;

exports[`unit:lib/toUrl > should return \`id\` as \`URL\` (3) 1`] = `"node:test"`;

exports[`unit:lib/toUrl > should return \`id\` as \`URL\` (4) 1`] = `file://\${process.cwd()}/tsconfig.build.json`;
